// zappy-db/main.js
var path = require('path');

function _getAppDataRoot() {
	switch (process.platform) {
	case 'win32':
	case 'win64':
		return path.join(process.env["LOCALAPPDATA"] || process.env["APPDATA"],
				"zappyglasses");
	case 'darwin':
		return path.join(process.env["HOME"], "Library", "Application Support",
				"zappyglasses");
	}
}

function DAO() {
	console.log("about to open the database...");
	var dbPath = path.join(_getAppDataRoot(), 'nedb')
	console.log('nedb path = ' + dbPath);
	var fs = require('fs')
	if (!fs.existsSync(dbPath)) {
		fs.mkdirSync(dbPath, '0755');
	}
	var Datastore = require('nedb');
	this.db = {};
	this.db.profiles = new Datastore({
		filename : path.join(dbPath, 'profiles.nedb')
	});
	this.db.files = new Datastore({
		filename : path.join(dbPath, 'files.nedb')
	});
	this.db.profiles.loadDatabase(function(err) {
		if (err !== null) {
			console.log('Failed to load profiles database: ' + err);
		}
	});
	this.db.files.loadDatabase(function(err) {
		if (err !== null) {
			console.log('Failed to load files database: ' + err);
		}
	});
}

DAO.prototype.getFilePath = function(id, callback) {
	this.db.files.find({
		_id : id
	}, function(err, docs) {
		if (err !== null) {
			callback(0, undefined);
		} else {
			callback(1, docs[0].path);
			console
					.log("Data found for Id = " + id + " file = "
							+ docs[0].path);
		}
	});
}

DAO.prototype.saveFilePath = function(path, callback) {
	var filesDB = this.db.files;
	filesDB.find({
		path : path
	}, function(err, docs) {
		if (err !== null) {
			callback(0, undefined);
		} else {
			if (docs === null || docs.length === 0) {
				var doc = {};
				doc.path = path;
				doc.usedTime = new Array();
				doc.usedTime.push(new Date().getTime());
				filesDB.insert(doc, function(err, newDoc) {
					if (err !== null) {
						callback(0, undefined);
					} else {
						callback(1, newDoc._id);
						console.log("File added = " + newDoc);
					}
				});
			} else {
				var doc = docs[0];
				doc.usedTime.push(new Date().getTime());

				filesDB.update({
					_id : doc._id
				}, doc, {}, function(err, numReplaced) {
					if (err !== null) {
						callback(0, undefined);
					} else {
						callback(1, doc._id);
						console.log("Updated row (" + numReplaced + ") = "
								+ doc._id);
					}
				});
			}
		}
	});
}

DAO.prototype.getZappyGlass = function(id, callback) {
	this.db.profiles.find({
		_id : id
	}, function(err, docs) {
		if (err !== null) {
			callback(0, undefined);
		} else {
			callback(1, docs[0]);
			console.log("Found for Id = " + id + " glass = " + docs[0]);
		}
	});
}

DAO.prototype.getAllZappyGlasses = function(callback) {
	this.db.profiles.find({}, function(err, docs) {
		if (err !== null) {
			callback(0, undefined);
		} else {
			callback(1, docs);
			console.log("All glasses = " + docs);
		}
	});
}

DAO.prototype.saveZappyGlass = function(glass, callback) {

	this.db.profiles.insert(glass, function(err, newDoc) {
		if (err !== null) {
			callback(0, undefined);
		} else {
			callback(1, newDoc);
			console.log("Glass inserted = " + newDoc);
		}
	});
}

var dao = new DAO();

dao.DAO = DAO;
module.exports = dao;
